### PR TITLE
add Surplus

### PR DESCRIPTION
**Pull request by adamhaile**
Add [Surplus](https://github.com/adamhaile/surplus) to the list of frameworks.

I think I figured out all the places that needed editing to integrate another framework, but if not, let me know and I'm happy to revise.

As a framework author, this project is really useful for testing web component interop, so thanks for it!

### CODE REVIEW

**Comment by robdodson** [comment]
Hey sorry for the delay. I'm on vacation till the 18th. Will definitely review then. Sorry for the wait!

**Comment by brodybits** [comment]
Any progress?

**Comment by robdodson** [comment]
Hey! I was literally just staring at this email when your reply came in. Yes, I'm going to try to get to this today now that I'm back and (mostly) caught up on email.

**Comment by robdodson** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by adamhaile** [comment]
Recognizing that Surplus isn't the best known framework, if there's any help I can give you, let me know.  If it matters, I started from the React example and modified it to apply to Surplus.

File: libraries/surplus/src/advanced-tests.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  ComponentWithChildrenRerender,
+  ComponentWithDifferentViews,
+  ComponentWithProperties,
+  ComponentWithUnregistered,
+  ComponentWithImperativeEvent,
+  ComponentWithDeclarativeEvent
+} from "./components";
+
+
+//let _it = it;
```
Comment (robdodson) [inline libraries/surplus/src/advanced-tests.js]: Did you mean to leave this comment?

File: libraries/surplus/src/basic-tests.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  ComponentWithChildren,
+  ComponentWithChildrenRerender,
+  ComponentWithDifferentViews,
+  ComponentWithProperties,
+  ComponentWithUnregistered,
+  ComponentWithImperativeEvent,
+  ComponentWithDeclarativeEvent
+} from "./components";
+
+//let _it = it;
```
Comment (robdodson) [inline libraries/surplus/src/basic-tests.js]: same as above, did you mean for this comment to stick around?

File: libraries/surplus/src/advanced-tests.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+//it = function it(title, fn) { return _it.call(this, title, () => S.root(fn.bind(this))); }
+
+describe("advanced support", function() {
+
+  describe("attributes and properties", function() {
+    it("will pass array data as a property", function() {
+      S.root(() => {
+        this.weight = 2;
```
Comment (robdodson) [inline libraries/surplus/src/advanced-tests.js]: for consistency would you mind moving all calls to `this.weight` out of this function.

File: libraries/surplus/src/basic-tests.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+      });
+    });
+
+    it("will pass numeric data as either an attribute or a property", function() {
+      S.root(() => {
+        this.weight = 3;
+        let root = <ComponentWithProperties />;
+        let wc = root.wc;
+        let data = wc.num || wc.getAttribute("num");
+        expect(data).toEqual(42);
```
Comment (robdodson) [inline libraries/surplus/src/basic-tests.js]: I know it's super annoying, but recently we switched from expect.js to chai for expectations.

For example, you can see how we're handling a similar test case in react now: https://github.com/webcomponents/custom-elements-everywhere/blob/master/libraries/react/src/basic-tests.js#L118

I think the PR will need to be updated to use chai.

File: libraries/surplus/src/components.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+export const ComponentWithChildren = () =>
+    <div>
+        <ce-with-children ref={__.wc}></ce-with-children>
+    </div>;
+
+export const ComponentWithChildrenRerender = () => {
+    const count = S.data(1);
+    Promise.resolve().then(() => count(count() + 1));
+    //S.cleanup(() =>  clearInterval(this.interval)); // what does this do? where does .interval come from?
```
Comment (robdodson) [inline libraries/surplus/src/components.js]: ?

**Comment by robdodson** [comment]
OK, added some feedback. Thanks for working on this @adamhaile !

File: libraries/surplus/src/components.js (hunk 5 above)
Comment (adamhaile) [reply libraries/surplus/src/components.js]: ? here as well :).  I started with the React impl, and it has this line: https://github.com/webcomponents/custom-elements-everywhere/blob/50fb30f0e8200ddb0743a3609a20636f998edfb3/libraries/react/src/components.js#L54 .  I couldn't see anywhere in the codebase that this.interval was being set?  Is that something the test harness uses?

File: libraries/surplus/src/advanced-tests.js (hunk 1 above)
Comment (adamhaile) [reply libraries/surplus/src/advanced-tests.js]: nope, will remove.  Was left from trying to see if I could wrap calls to `it()` in `S.root()` so as to modify the test themselves less.

File: libraries/surplus/src/advanced-tests.js (hunk 3 above)
Comment (adamhaile) [reply libraries/surplus/src/advanced-tests.js]: Ah!  This explains why my attempt to wrap `it()` didn't work -- the weight calculations weren't passing through.  Will fix.

File: libraries/surplus/src/basic-tests.js (hunk 4 above)
Comment (adamhaile) [reply libraries/surplus/src/basic-tests.js]: Sure, can do.  Is https://github.com/webcomponents/custom-elements-everywhere/commit/e294e8d150ba02ea57f6e5119056e5f09f92d9e5#diff-aca65ac5cb69fb4b776030c3d7659d70 a good reference for the changes?

File: libraries/surplus/src/components.js (hunk 5 above)
Comment (brodybits) [reply libraries/surplus/src/components.js]: Wondering if that should be removed from the React version?

**Comment by adamhaile** [comment]
updated per review.  also bumped surplus version to release, now that it's out.

**Comment by robdodson** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by robdodson** [comment]
I just fixed up the branch and merged it into master. I think it should but up on the site now. Thank you!

**Pull request closed without merging**